feat(initial): initial work

SF2311 opened the pull request.

SF2311 committed:
feat(initial): initial work

SF2311 committed:
initial

haslersn committed:
fix: set default for postfix_lookup_table_configs

SF2311 committed:
feat(opendkim): support verifying only

haslersn commented on files/saslauthd-postfix:
```diff
@@ -0,0 +1,44 @@
+# {{ ansible_managed }}
```
This file is not a template, it's just copied to the host.

SF2311 replied:
I removed the line.

neumantm replied:
Why remove?
The comment is still useful.
I'd suggest just hardcoding the message. 
Alternatively: Template that file jsut for this (we did this before)

haslersn replied:
Done

haslersn commented on files/saslauthd-postfix:
```diff
@@ -0,0 +1,44 @@
+
+# Short name of this saslauthd instance. Strongly recommended.
+# (suggestion: saslauthd)
+NAME="saslauthd-pstfx"
```
Why abbreviate "postfix"?

SF2311 replied:
Because the name can be 15 characters at most.

haslersn commented on templates/saslauthd.conf.j2:
```diff
@@ -0,0 +1,9 @@
+ldap_bind_pw: {{ postfix_bind_pw }}
+{% endif %}
+ldap_search_base: {{ postfix_ldap_search_base }}
+ldap_auth_method: bind
```
Why don't we just put all of these config options in a single dict and place every key-value pair of that dict into this file?

SF2311 replied:
Done.

haslersn commented on tasks/main.yml:
```diff
@@ -1,2 +1,111 @@
+        groups: sasl
+
+  when: postfix_ldap_servers is defined
+  notify: Restart saslauthd service
```
Maybe you should also stop and disable the service when `postfix_ldap_servers` is not defined.

SF2311 replied:
Done.

haslersn commented on tasks/main.yml:
```diff
@@ -1,2 +1,111 @@
+      copy:
+        content: "{{ postfix_opendkim_config }}"
+        dest: /etc/opendkim.conf
+      notify: Restart opendkim service
```
We should probably notify this handler whenever anything in this block is changed.

Also stop and disable the service when `postfix_opendkim_enabled` is not defined.

SF2311 replied:
Done.

haslersn commented on tasks/main.yml:
```diff
@@ -1,2 +1,111 @@
+  copy:
+    content: "{{item.content}}"
+    dest: "/etc/postfix/{{item.name}}.cf"
+  loop: "{{ postfix_lookup_table_configs }}"
```
We should probably rename this variable. And why don't we place `main.cf` and `master.cf` using the same variable? I.e.:

```yaml
postfix_lookup_table_configs:
  main: |
    ...
  master: |
    ...
```

SF2311 replied:
Done.

SF2311 committed:
feat(review): address points from review

SF2311 committed:
feat(tasks): fix error msg

SF2311 committed:
feat(sasl): explicit option to enable user authentication

SF2311 committed:
feat(tasks): don't log config file content

haslersn committed:
feat(opendkim): support placing keys

haslersn committed:
feat(opendkim): support placing additional configs

haslersn committed:
chore(opendkim): remove unnecessary permissions from config directories

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,209 @@
+
+- name: Disable and stop saslauthd
+  block:
+    - name: Install python-apt
```
This seems like a lot of overhead just to disable and stop a service.

neumantm replied:
This scenario, where the service is installed and running and then needs to be disabled and stopped, will be very rare.
Do we really need this feature, or can we just document that this is missing?

neumantm replied:
Alernative implementations (which are not perfect either):

- First mask the service (this can be done even if it does not exist)
  - Then it can also be stopped and disabled
  - However, then there is this fake service even on machines where this feature was never enabled.
- Instead of  using the list of packages use the list of services (`systemctl list-units`)

haslersn replied:
> This scenario, where the service is installed and running and then needs to be disabled and stopped, will be very rare.

We already ran into this scenario on mail-out01.

haslersn replied:
Fixed it using another way:

```yaml
  register: postfix_service_stop
  failed_when: postfix_service_stop is failed and 'Could not find the requested service ' not in postfix_service_stop.msg
```

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,209 @@
+
+    - name: Place opendkim config
+      copy:
+        content: "{{ postfix_opendkim_config }}"
```
I think we should either mark this as `required` or provide a default for it.

haslersn replied:
Do you mean in the readme or in the template (`| mandatory`)? The latter is not necessary, because in this context ansible throws an error when an undefined variable is used in a template.

> The task includes an option with an undefined variable. The error was: 'postfix_opendkim_config' is undefined. 'postfix_opendkim_config' is undefined

neumantm replied:
Ok. I did not know that.

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,209 @@
+        content: "{{ item.value }}"
+        dest: /etc/opendkim/{{ item.key }}.cf
+      loop: "{{ postfix_opendkim_additional_configs | dict2items }}"
+      no_log: True
```
Why?

haslersn replied:
Might contain database passwords.

haslersn replied:
Though I now replaced `no_log: True` by `loop_control.label: "{{ item.key }}"`.

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,209 @@
+        state: absent
+        path: "{{ path }}"
+      loop: >-
+        {{postfix_opendkim_confs_to_delete | map('regex_replace', '^', '/etc/opendkim/') | map('regex_replace', '$', '.cf')}}
```
Instead of the mapping this could just be done when templating the path inside the loop...

haslersn replied:
Done

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,209 @@
+        append: True
+        groups: sasl
+
+  when: postfix_sasl_enabled is defined
```
Why work with is defined here and not with boolean values and a default of `false`?

(Same for other `_enabled` variables)

haslersn replied:
Fixed.

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,209 @@
+      {{
+        postfix_existing_config_files.files
+        | map(attribute='path')
+        | select("match", "^/etc/postfix/.*\.cf$")
```
Which file would be returned by the above find which is not matched here?

haslersn replied:
I removed the select.

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,209 @@
+  notify:
+    - Restart postfix service
+  loop: >-
+    {{postfix_confs_to_delete | map('regex_replace', '^', '/etc/postfix/') | map('regex_replace', '$', '.cf')}}
```
Same as above.

haslersn replied:
Done

neumantm commented on files/smtpd.conf:
Maybe we should name this file `sasl-smtpd.conf` or something?

haslersn replied:
Done

neumantm reviewed (changes requested):
Please update `README.md` and `meta/main.yml`

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,209 @@
+  when: postfix_opendkim_enabled is defined
+  notify: Restart opendkim service
+
+- name: Disable and stop saslauthd
```
```suggestion
- name: Disable and stop opendkim
```

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,209 @@
+      package_facts:
+        manager: auto
+
+    - name: Stop and disable saslauthd service
```
```suggestion
    - name: Stop and disable opendkim service
```

neumantm commented:
Did not check whether this actually the correct (and complete) procedure to set up a postfix server with sasl and dkim (yet).

haslersn committed:
Update tasks/main.yml

Co-authored-by: Tim Neumann <neumantm@fius.informatik.uni-stuttgart.de>

haslersn committed:
simplify tasks for deletion of config files

haslersn committed:
set loop label instead of no_log

haslersn committed:
Update tasks/main.yml

Co-authored-by: Tim Neumann <neumantm@fius.informatik.uni-stuttgart.de>

haslersn committed:
use booleans for *_enabled variables

haslersn committed:
add ansible_managed to saslauthd-postfix

haslersn committed:
rename smtpd.conf -> sasl-smtpd.conf

haslersn committed:
Less overhead for stopping and disabling services

SF2311 committed:
chore(defaults): remove unused variable

SF2311 committed:
feat(config): don't delete main.cf and master.cf

If the default configuration should be used, we cannot delete these
files even though they might not be managed by ansible.

SF2311 committed:
wip(README): WIP

haslersn committed:
refactor: rename postfix_additional_configs -> postfix_configs

haslersn committed:
chore(defaults): sort defaults/main.yml

haslersn committed:
docs: rework README

haslersn committed:
docs: fill out meta/main.yml

neumantm commented on README.md:
````diff
@@ -1,42 +1,144 @@
+This dict will be parsed and written to the saslauthd config file.
+The default is:
 
 ```yml
+postfix_saslauthd_options:
+  ldap_servers: "{{ postfix_ldap_servers }}"
+  ldap_search_base: "{{ postfix_ldap_search_base }}"
+  ldap_auth_method: bind
 ```
````
I'd prefer if we had a way to make sure the actual default does not diverge from this.
Either just reference the `defaults/main.yml` (as already done above) or place a comment in `defaults/main.yml` reminding editors to change the README, too.

haslersn replied:
I removed the default and phrased it as an example instead. There was no reason to have these `postfix_ldap_*` variables, because they were only used in this default. Instead one can directly configure `postfix_saslauthd_options`.

neumantm commented on README.md:
````diff
@@ -1,42 +1,144 @@
+
+The default config shown below, sets OpenDKIM up in verifier mode, listening at `localhost:12301`:
+
+```
+Syslog                  yes
+SyslogSuccess           yes
+Mode                    v
+
+# In Debian, opendkim runs as user "opendkim". A umask of 007 is required when
+# using a local socket with MTAs that access the socket as a non-privileged
+# user (for example, Postfix). You may need to add user "postfix" to group
+# "opendkim" in that case.
+UserID                  opendkim
+UMask                   007
+
+# Socket for the MTA connection (required).
+Socket                  inet:12301@localhost
+PidFile                 /run/opendkim/opendkim.pid
+
+# The trust anchor enables DNSSEC. In Debian, the trust anchor file is provided
+# by the package dns-root-data.
+TrustAnchorFile         /usr/share/dns/root.key
+```
````
Same as above

haslersn replied:
Done

neumantm commented on defaults/main.yml:
```diff
@@ -0,0 +1,30 @@
+  PidFile                 /run/opendkim/opendkim.pid
+
+  # The trust anchor enables DNSSEC. In Debian, the trust anchor file is provided
+  # by the package dns-root-data.
```
Maybe add as a note: Is this package automatically installed with the role (via dependencies) or do admins need to install this through another means? If the latter, a comment in the README stating this would also be apropriate I guess.

SF2311 replied:
It is automatically installed as a dependency.

neumantm commented on README.md:
```diff
@@ -1,42 +1,144 @@
+### DKIM Private Keys
+
+The `postfix_opendkim_keys` variable can be used to copy DKIM private keys to the host.
+This variable takes is a dict where each key-value pair describes a private key.
```
Not an english sentence.

haslersn replied:
Fixed

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,184 @@
+  package:
+    name:
+      - postfix
+      - postfix-ldap
+      - postfix-pgsql
```
Maybe we could add to the README, that we also install those.

SF2311 replied:
Done.

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,184 @@
+        dest: /etc/default/saslauthd-postfix
+
+    - name: Create socket directory for saslauthd
+      shell: dpkg-statoverride --add root sasl 755 /var/spool/postfix/var/run/saslauthd
```
Is `755`, necesarry here?
The [debian docs](https://wiki.debian.org/PostfixAndSASL) make it `710`.

SF2311 replied:
These instructions are for Debian 7. I tried it with 710 and postfix didn't authenticate users with a permission denied error.
In theory 750 should be enough, but this also did not work. I hadn't had the time to investigate further at the time, so I set it to 755.

neumantm commented on tasks/main.yml:
```diff
@@ -1,2 +1,184 @@
+        append: True
+        groups: opendkim
+
+    - name: Place opendkim keys
```
How does opendkim find these keys?
Is it a magic folder?
Based on what I can find online the config would need a `KeyFile` or `KeyTable` config.

haslersn replied:
Yes, one needs to reference them in the config. That's the same as with additional files placed via `postfix_configs`.

neumantm reviewed (changes requested):
Otherwise looks good.

haslersn committed:
remove postfix_ldap_* variables, instead one can use postfix_saslauthd_options

haslersn committed:
README: refer to defaults/main.yml for postfix_opendkim_config

haslersn committed:
README: fix typo

SF2311 committed:
chore(doc): address comments from review
